Doctrine DBAL store

Miliooo opened the pull request:
I needed a key value store dbal implementation for my own project. 

The only thing I found was https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DBALStorage.php and this package.

So I decided to implement one that implements this projects interface.

webmozart commented:
Hi @Miliooo, thanks a lot for this! Could you fix the StyleCI issues please?

Did you fix the failing test you had problems with?

jdreesen commented on composer.json:
```diff
@@ -17,12 +17,14 @@
         "basho/riak": "To enable the RiakStore",
-        "doctrine/cache": "To enable the CachedStore"
+        "doctrine/cache": "To enable the CachedStore",
+        "doctrine/dbal": "To enable the Dbalstore"
```
Store with a capital s?

Miliooo replied:
thx fixed

Miliooo commented:
Hi @webmozart, I updated the StyleCI config and also applied their patch.
Yeah I fixed the failing test, it was a logical error on my part. 

Returning null from the query, which is the same as the default value, so in my logic, the key did not exist. That's why the code now returns the database row.

I guess this is ready for review.

tgalopin commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+    public function clear()
+    {
+        try {
+            $connection = $this->connection->query('DELETE FROM '.$this->table);
```
I would name this `$query`.

Miliooo replied:
I guess the right name would be statement since the method returns a statement object. Hope I'm not the only one who finds this very confusing

webmozart replied:
Yes, let's call this `$stmt`.

webmozart commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+     */
+    public function __construct(Connection $connection, $table = 'store')
+    {
+        Assert::string($table, 'The table must be a string. Got: %s');
```
You can use the combined constrained `stringNotEmpty()` here.

tgalopin commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+        return $result;
+    }
+
+    public function configureTable()
```
Can you add a phpdoc comment here as it's a public method?

Miliooo replied:
Guess I'll make it protected, since the entry point is the method configureSchema.

webmozart replied:
I'd keep this as a public method, rename to `createTable()` and drop `configureSchema()` instead (as explained below).

webmozart commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+
+        try {
+            $existing = $this->exists($key);
+        } catch (\Exception $e) {
```
Can you change this to `(Exception $e)` and add `use Exception;` at the top?

webmozart commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+    {
+        try {
+            $stmt = $this->connection->query('SELECT meta_key FROM '.$this->table);
+            $result = $stmt->fetchAll(\PDO::FETCH_COLUMN);
```
same here: `(PDO::...`, `use PDO;`

webmozart commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+     *
+     * @return \Doctrine\DBAL\Schema\Table|null
+     */
+    public function configureSchema(Schema $schema)
```
Why do we need this method?

Miliooo replied:
Well I 'borrowed' that code from [broadway event store] (https://github.com/qandidate-labs/broadway/blob/master/src/Broadway/EventStore/DBALEventStore.php#L146)

I used it to setup my tests, and it makes it possible to do add a [create database command] (https://github.com/qandidate-labs/broadway/blob/master/src/Broadway/Bundle/BroadwayBundle/Command/SchemaEventStoreCreateCommand.php) in symfony. Maybe in the readme we could provide a database schema too... Or maybe add a reference to that function so users can see how they have to create their database.

webmozart replied:
I see. I'd drop this method and add a public `getTableName()` then. Hiding the `hasTable()` statement here and adding a `null` check instead in the calling code seems a bit magical to me.

I'd also rename the `$table` property to `$tableName` to be clear that this is a string.

webmozart commented on src/DbalStore.php:
```diff
@@ -0,0 +1,303 @@
+        $table->setPrimaryKey(array('id'));
+        $table->addUniqueIndex(array('meta_key'));
+
+        return $table;
```
Why does this method return a value? How is this method used?

Miliooo replied:
It's used for setting up my tests, and can be used to let doctrine setup your tests, see my previous comment and https://github.com/webmozart/key-value-store/pull/15/files#diff-840e95275d34986f00b5591a212cbcd2R39

webmozart replied:
Ok :) Let's rename it to `createTable()` then?

Miliooo replied:
Well it doesn't really create a table, it's an object representation of a table. 
The schema manager already has the method `createTable.` Maybe `getTableSchema`.
This gives the following code 
```
        $tableName = self::$dbalStore->getTableName();
        if ($schema->hasTable($tableName)) {
            return;
        }

        $schemaManager->createTable(self::$dbalStore->getTableSchema());
```

I like it more then two createTableCalls on different objects. It also returns a Table object, so I like it more then getSchemaTable (which would suggest we return a schema object)

webmozart replied:
I see. `getTableForCreate()` maybe?

webmozart replied:
```php
if (!$schema->hasTable(self::$dbalStore->getTableName())) {
    $schemaManager->createTable(self::$dbalStore->getTableForCreate());
}
```

Miliooo replied:
cool, yeah no need to add a temporary var

webmozart commented:
Looks good! Can you update the README and CHANGELOG as well?

Miliooo commented:
Did a bunch of updates, I'll squash some commits later. If there are other changes needed let me know.

webmozart commented on src/DbalStore.php:
```diff
@@ -0,0 +1,305 @@
+    /**
+     * Object Representation of the table used in this class.
+     *
+     * @return \Doctrine\DBAL\Schema\Table
```
You can shorten this to `@return Table` and add `use Doctrine\DBAL\Schema\Table` at the top. :)

Miliooo replied:
Yes, but you have to know your abc very well because it also has to be in the right order :mortar_board: 
https://www.youtube.com/watch?v=75p-N9YKqNo

Miliooo commented:
I screwed up my git commits. Not sure if you are ok with 6 different commits. I could resubmit a new pull request. Thing that got me stuck are those ```merge branch``` messages. ```git rebase -i HEAD~``` doesn't seem to work very well on those.

Been a while since I submitted pull requests, I'm just pushing to master branch all the time on my personal project.

Miliooo commented:
closed in favor of https://github.com/webmozart/key-value-store/pull/17
